src/mds/ScrubStack.cc

```diff
@@ -159,156 +155,74 @@ void ScrubStack::kick_off_scrubs()
-	// we added new stuff to top of stack, so reset ourselves there
-	it = scrub_stack.begin();
-      } else {
-	dout(20) << __func__ << " dirfrag, no progress" << dendl;
-	can_continue = false;
       }
     } else {
       ceph_assert(0 == "dentry in scrub stack");
     }
   }
 }
 
-void ScrubStack::scrub_dir_inode(CInode *in,
-                                 bool *added_children,
-                                 bool *done)
+void ScrubStack::scrub_dir_inode(CInode *in, bool *added_children, bool *done)
 {
   dout(10) << __func__ << " " << *in << dendl;
 
-  *added_children = false;
-  bool all_frags_done = true;
-
   ScrubHeaderRef header = in->get_scrub_header();
-  ceph_assert(header != nullptr);
+  ceph_assert(header);
 
-  if (header->get_recursive()) {
-    frag_vec_t scrubbing_frags;
-    std::queue<CDir*> scrubbing_cdirs;
-    in->scrub_dirfrags_scrubbing(&scrubbing_frags);
-    dout(20) << __func__ << " iterating over " << scrubbing_frags.size()
-      << " scrubbing frags" << dendl;
-    for (const auto& fg : scrubbing_frags) {
-      // turn frags into CDir *
-      CDir *dir = in->get_dirfrag(fg);
-      if (dir) {
-	scrubbing_cdirs.push(dir);
-	dout(25) << __func__ << " got CDir " << *dir << " presently scrubbing" << dendl;
-      } else {
-	in->scrub_dirfrag_finished(fg);
-	dout(25) << __func__ << " missing dirfrag " << fg << " skip scrubbing" << dendl;
-      }
-    }
+  MDSGatherBuilder gather(g_ceph_context);
 
-    dout(20) << __func__ << " consuming from " << scrubbing_cdirs.size()
-	     << " scrubbing cdirs" << dendl;
+  frag_vec_t frags;
+  in->dirfragtree.get_leaves(frags);
+  dout(20) << __func__ << "recursive mode, frags " << frags << dendl;
 
-    while (g_conf()->mds_max_scrub_ops_in_progress > scrubs_in_progress) {
-      // select next CDir
-      CDir *cur_dir = NULL;
-      if (!scrubbing_cdirs.empty()) {
-	cur_dir = scrubbing_cdirs.front();
-        scrubbing_cdirs.pop();
-	dout(20) << __func__ << " got cur_dir = " << *cur_dir << dendl;
-      } else {
-	bool ready = get_next_cdir(in, &cur_dir);
-	dout(20) << __func__ << " get_next_cdir ready=" << ready << dendl;
-
-	if (ready && cur_dir) {
-	  cur_dir->scrub_initialize(header, nullptr);
-	  scrubbing_cdirs.push(cur_dir);
-	} else if (!ready) {
-	  // We are waiting for load of a frag
-	  all_frags_done = false;
-	  break;
-	} else {
-	  // Finished with all frags
-	  break;
-	}
-      }
-      // scrub that CDir
-      bool frag_added_children = false;
-      bool frag_done = false;
-      scrub_dirfrag(cur_dir,
-		    &frag_added_children, &frag_done);
-      if (frag_done) {
-	cur_dir->inode->scrub_dirfrag_finished(cur_dir->frag);
-      }
-      *added_children |= frag_added_children;
-      all_frags_done = all_frags_done && frag_done;
-    }
-
-    dout(20) << "finished looping, all_frags_done=" << all_frags_done << dendl;
-  } else {
-    dout(20) << "!scrub_recursive" << dendl;
+  for (auto &fg : frags) {
```

Comment:
@ukernel This looks like a strict breadth first search approach. Why is it given as pre-order traversal in the commit message?

Reply:
This loop just add objects to The queue.  The order I mean is the order of executing scrub_dir_inode_final().  https://www.geeksforgeeks.org/tree-traversals-inorder-preorder-and-postorder/

Reply:
I think the meant the same.
```
                        1(inode)
                     /           \
            2(dirfrag)      3 (dirfrag)
           /         \
   4(inode)       5(inode)

```

If this is the directory hierarchy then isn't the order of operations going to be: 
1) Start scrubbing 1 -> add 2 and 3 to the queue -> do final scrub of 1/finish scrubbing 1. (Contents of the Queue: 2,3)
2) Start scrubbing 2 -> add 4 and  5 to the queue -> do final scrub of 2/finish scrubbing 2. (Contents of the Queue: 3,4,5).
.
.

So isn't the order of scrubbing going to 12345 (BFS) rather than 12453 (Pre-order traversal)?

Reply:
Ok, I think you are right. will change the commit message later

Reply:
depth of tree in your example is too small. In following example. the order is 12453
```
                        1(inode)
                     /           \
            2(inode)      3(inode)
           /         \
   4(inode)       5(inode)

```

Reply:
I'm still confused on how the order is 12453 and not 12345 even in your example. If I consider the whole tree with dirfrags then it is:
```
                      1(inode)
                           |             
                     1(dirfrag)
                  /             \
           2(inode)           3(inode)
                |
           2(dirfrag)
         /           \
     4(inode)     5(inode)
```
In which case order of operations is going to be:
1) Start scrubbing 1(inode) -> add 1(dirfrag) to the queue -> do final scrub/finish scrubbing 1(inode). (Contents of the Queue: `[ 1(dirfrag) ]`)
2) Start scrubbing 1(dirfrag) -> add 2(inode) and 3(inode) to the queue -> do final scrub/finish scrubbing 1(dirfrag). (Contents of Queue: `[ 2(inode), 3(inode) ]`)
3) Start scrubbing 2(inode) -> add 2(dirfrag) to the queue -> do final scrub/finish scrubbing 2(inode). (Contents of the Queue: `[ 3(inode), 2(dirfrag) ]`)
4) Start scrubbing 3(inode) -> do final scrub/finish scrubbing 3(inode). (Contents of the Queue: `[ 2(dirfrag) ]`)
5) Start scrubbing 2(dirfrag) -> add 4(inode) and 5(inode) to the queue -> do final scrub/finish scrubbing 2(dirfrag). (Contents of the Queue: `[ 4(inode), 5(inode) ]`)
6) Scrub 4(inode)
7) Scrub 5(inode) 

So here if we look at the order of scrubbing the inodes it is still 12345(BFS).

Reply:
At step 4 . 2(dirfrag) is scrubbed instead of 3(inode). Because dirfrags are always queued at front.

Reply:
@ukernel Oh thanks! I missed that dirfrags get pushed to the top of the queue. But still, the inodes 4 and 5 will get pushed to the back of the queue (contents: `[ 3(inode) ]`) before finishing scrub of 2(dirfrag), so the order is going to be 3(inode), 4(inode) and 5(inode) after 2(dirfrag) right?

Reply:
I think you are right. I will change the commit message